tools/eslint-rules/prefer-common-expectserror.js

```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+const msg = 'Please use common.expectsError(fn, err) instead of ' +
+            'assert.throws(fn, common.expectsError(err)).';
+
+function isAssertThrows(node) {
+  return node &&
+         node.callee &&
+         node.callee.object &&
+         node.callee.object.name === 'assert' &&
+         node.callee.property &&
+         node.callee.property.name === 'throws';
+}
+
+
+function isFunction(arg) {
+  return arg && (arg.type === 'ArrowFunctionExpression' ||
+                 arg.type === 'FunctionExpression' ||
+                 arg.type === 'CallExpression');
+}
+
+function isCommonExpectsError(node) {
+  return node &&
+         node.callee &&
+         node.callee.object &&
+         node.callee.object.name === 'common' &&
+         node.callee.property &&
+         node.callee.property.name === 'expectsError';
+}
+
+module.exports = function(context) {
+  return {
+    CallExpression(node) {
```

Comment:
Could `no-restricted-syntax` handle this?

Reply:
Hmm, possibly but it'll definitely be a pretty convoluted rule in that case. There's no "or" for AST selectors, as far as I know, so it would get super verbose and repetitive to account for all these possibilities.

Reply:
You don't need the "or" condition since that checks for a function as the first argument to `assert.throws()`, which is always required. This selector seems to match all of the cases you have in this PR:

```
"CallExpression[callee.object.name='assert'][callee.property.name='throws'][arguments.length<3][arguments.1.type='CallExpression'][arguments.1.callee.object.name='common'][arguments.1.callee.property.name='expectsError']"
```

It's not beautiful, but it might be able to be simplified further.

Reply:
Good point. That should work. Will update.

Reply:
Another problem with `no-restricted-syntax` here is that this is a rule that should only apply in `test`. I'm not sure if there's a way to specify `no-restricted-syntax` in `test` in a way that won't override `no-restricted-syntax` from the config file in the parent directory. /cc @not-an-aardvark

Reply:
I'm guessing that's why we also have `prefer-common-mustnotcall` rule.

Reply:
Unfortunately, there's not a good way to do that at the moment. As a workaround, you could copy-paste the configuration from the parent directory, add the new entry, and put the resulting config in `test/.eslintrc.yaml`. This would have the desired effect, but it could cause the two configs to get out of sync if someone updates `.eslintrc.yaml` later on and forgets to update the copied version in `test/.eslintrc.yaml`.

There is some discussion on improving this in https://github.com/eslint/eslint/issues/9192.

Reply:
That's unfortunate. I'm not a fan of adding custom rules that we don't need, but keeping the two config files in sync seems like a worse idea.